devel/fortio/http.go

```diff
@@ -147,17 +161,486 @@ func NewClient(url string, numConnections int, compression bool) *Client {
 		url,
 		req,
 		&http.Client{
-			Timeout: 3 * time.Second,
+			Timeout: 3 * time.Second, // TODO: make configurable
 			Transport: &http.Transport{
 				MaxIdleConns:        numConnections,
 				MaxIdleConnsPerHost: numConnections,
 				DisableCompression:  !compression,
 				Dial: (&net.Dialer{
-					Timeout: 1 * time.Second,
+					Timeout: 4 * time.Second,
 				}).Dial,
-				TLSHandshakeTimeout: 1 * time.Second,
+				TLSHandshakeTimeout: 4 * time.Second,
+			},
+			// Lets us see the raw response instead of auto following redirects.
+			CheckRedirect: func(req *http.Request, via []*http.Request) error {
+				return http.ErrUseLastResponse
 			},
 		},
 	}
 	return &client
 }
+
+// BasicClient is a fast, lockfree single purpose http 1.0/1.1 client.
+type BasicClient struct {
+	buffer       []byte
+	req          []byte
+	dest         net.TCPAddr
+	socket       *net.TCPConn
+	size         int
+	code         int
+	errorCount   int
+	headerLen    int
+	url          string
+	host         string
+	hostname     string
+	port         string
+	http10       bool // http 1.0, simplest: no Host, forced no keepAlive, no parsing
+	keepAlive    bool
+	parseHeaders bool // don't bother in http/1.0
+}
+
+// NewBasicClient makes a basic, efficient http 1.0/1.1 client.
+// This function itself doesn't need to be super efficient as it is created at
+// the beginning and then reused many times.
+func NewBasicClient(urlStr string, proto string, keepAlive bool) Fetcher {
+	// Parse the url, extract components.
+	url, err := url.Parse(urlStr)
+	if err != nil {
+		Err("Bad url '%s' : %v", urlStr, err)
+		return nil
+	}
+	if url.Scheme != "http" {
+		Err("Only http is supported, can't use url %s", urlStr)
+		return nil
+	}
+	// note: Host includes the port
+	bc := BasicClient{url: urlStr, host: url.Host, hostname: url.Hostname(), port: url.Port(), http10: (proto == "1.0")}
+	bc.buffer = make([]byte, (*bufferSizeKbFlag)*1024)
+	if bc.port == "" {
+		bc.port = url.Scheme // ie http which turns into 80 later
+		LogV("No port specified, using %s", bc.port)
+	}
+	addrs, err := net.LookupIP(bc.hostname)
+	if err != nil {
+		Err("Unable to lookup '%s' : %v", bc.host, err)
+		return nil
+	}
+	if len(addrs) > 1 && DbgOn() {
+		Dbg("Using only the first of the addresses for %s : %v", bc.host, addrs)
+	}
+	Dbg("Will go to %s", addrs[0])
+	bc.dest.IP = addrs[0]
+	bc.dest.Port, err = net.LookupPort("tcp", bc.port)
+	if err != nil {
+		Err("Unable to resolve port '%s' : %v", bc.port, err)
+		return nil
+	}
+	// Create the bytes for the request:
+	host := bc.host
+	if hostOverride != "" {
+		host = hostOverride
+	}
+	bc.req = []byte("GET " + url.RequestURI() + " HTTP/" + proto + "\r\n")
+	if !bc.http10 {
+		bc.req = append(bc.req, []byte("Host: "+host+"\r\n")...)
+		bc.parseHeaders = true
+		if keepAlive {
+			bc.keepAlive = true
+		} else {
+			bc.req = append(bc.req, []byte("Connection: close\r\n")...)
+		}
+	}
+	for h := range extraHeaders {
```

Comment:
did you look at `bytes.Buffer` for usage here?  Instead of all the casting and appends, etc., you could have `buf.Write()` and `buf.WriteString()` calls ?